scripts/machine_translation/dataprocessor.py

```diff
@@ -216,33 +216,30 @@ def get_dataloader(data_set, args, dataset_type,
     else:
         raise NotImplementedError
 
+    data_lengths = get_data_lengths(data_set)
+
     if dataset_type == 'train':
-        data_lengths = get_data_lengths(data_set)
         train_batchify_fn = btf.Tuple(btf.Pad(), btf.Pad(),
                                       btf.Stack(dtype='float32'), btf.Stack(dtype='float32'))
 
-        batch_sampler = nlp.data.FixedBucketSampler(lengths=data_lengths,
-                                                    batch_size=args.batch_size,
-                                                    num_buckets=args.num_buckets,
-                                                    ratio=args.bucket_ratio,
-                                                    shuffle=True,
-                                                    use_average_length=use_average_length,
-                                                    num_shards=num_shards,
-                                                    bucket_scheme=bucket_scheme)
     else:
-        data_lengths = get_data_lengths(data_set)
-        target_lengths = list(map(lambda x: x[-1], data_lengths))
+        data_lengths = list(map(lambda x: x[-1], data_lengths))
         test_batchify_fn = btf.Tuple(btf.Pad(), btf.Pad(),
                                      btf.Stack(dtype='float32'), btf.Stack(dtype='float32'),
                                      btf.Stack())
 
-        batch_sampler = nlp.data.FixedBucketSampler(lengths=target_lengths,
-                                                    batch_size=args.test_batch_size,
-                                                    num_buckets=args.num_buckets,
-                                                    ratio=args.bucket_ratio,
-                                                    shuffle=False,
-                                                    use_average_length=use_average_length,
-                                                    bucket_scheme=bucket_scheme)
+    batch_sampler = nlp.data.FixedBucketSampler(lengths=data_lengths,
+                                                batch_size=(args.batch_size \
+                                                            if dataset_type == 'train' \
+                                                            else args.test_batch_size),
+                                                num_buckets=args.num_buckets,
+                                                ratio=args.bucket_ratio,
+                                                shuffle=(dataset_type == 'train'),
+                                                use_average_length=use_average_length,
+                                                num_shards=(num_shards \
+                                                            if dataset_type == 'train' \
+                                                            else 0),
```

Comment:
if else can be removed for `num_shards`, as the default value is 0. Though our current script only supports evaluation on one GPU, it can be extended to multi-GPU setting.

Reply:
thanks, and i think the `is else` code here is for the differ configuration of train dataset and val test dataset, remove this may not distinguish these. Do i understand right?

Reply:
In `make_dataloader` fn, you have already leave `num_shards` as it is in default for val & test sets. The if else here prevents the use of multi-gpu (num_shards > 0) for evaluation, so it is better to remove it.